Moving stuff over from core.

asynts opened the pull request:
Closes #1.

misha130 commented on src/Application/Extensions/IServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,17 @@
+using Microsoft.Extensions.DependencyInjection;
```
The class of the file should correspond to the name of the file.
Better just do it like in the Core where its a file called DependencyInjection.cs for each project in the root of it rather than a folder.
This is essentially a module of the project

asynts replied:
> This is essentially a module of the project

Good point. I put it into that folder because I thought that we would have a lot more extensions, but there is only really one that is responsible for dependency injection and it seems to be a common convention.

0ad01fb

misha130 commented on src/Application/Services/InitializationService.cs:
```diff
@@ -0,0 +1,45 @@
+namespace Codidact.Authentication.Application.Services
+{
+    /// <summary>
+    /// When the application is run for the first thme, there are a number of things
```
Typo

asynts replied:
615f0f6e

misha130 commented on src/Application/Services/InitializationService.cs:
```diff
@@ -0,0 +1,45 @@
+using System.Linq;
```
How about AdminService rather than InitalizationService? The methods don'l really relay the fact that its purpose is to initialize.

asynts replied:
I took the extra step with a service to have a collection point for all the initialization stuff. Currently, there is only the administrator stuff but I imagine there would be a lot more in the future, all sort of configuration that has to be done when the application is launched for the first time.

If you ever went though a Wordpress installation, that's what this if for. Stuff like:

  - Configuring the site name.
  - Choosing the subdomain or URL schema.
  - Configuring search engine stuff like `robots.txt`.
  - Adding API keys, e.g. Google Analytics.
  - etc.

I think most of this shouldn't be MVP but if I added the admin stuff there because having it hard coded would be dangerous if we don't remove it properly or if it's retained in the database for some reason. (I think there was a discussion on Discord about this, originally I had it hard-coded.)

I can name it `AdminService` or `AdministratorService` but that would likely be renamed. On the other hand, it would have to be rewritten anyways.

Do you think I should rename it?

asynts replied:
I decided to remove this entirely. This first-use page is something that should not go into the minimum viable product.

I re-added something similar to the original code that just creates an 'admin@codidact' user. This can be removed as soon as we have other means to create users, this would be done in #2.

fb47bd82

misha130 commented on src/Infrastructure/Extensions/IServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,42 @@
+        public static IServiceCollection AddInfrastructure(this IServiceCollection services, IConfiguration configuration)
+        {
+            services.AddDbContext<ApplicationDbContext>(options =>
+                options.UseInMemoryDatabase("authentication"));
```
In memory database? Why?

misha130 commented on src/WebApp/Pages/Account/Login.cshtml.cs:
```diff
@@ -0,0 +1,68 @@
+
+        public bool RememberLogin { get; set; } = false;
+
+        public string ReturnUrl { get; set; } = "/index";
```
Is this received from the request?

asynts replied:
I really thought that ASP.NET would figure this out somehow.

813dca6e

misha130 commented on src/WebApp/Pages/Account/Logout.cshtml.cs:
```diff
@@ -0,0 +1,43 @@
+            _events = events;
+        }
+
+        public string ReturnUrl { get; set; } = "/index";
```
Same as Login. Is this received from the logout request?

asynts replied:
813dca6

misha130 commented on src/WebApp/Startup.cs:
```diff
@@ -0,0 +1,55 @@
+        }
+
+        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
+        {
```
Missing HTTPS redirects, no?

asynts replied:
Yes, I removed that.

Since we split the project into multiple projects we will need a web-server like Nginx to divide the traffic and send the PII stuff to the authentication server and the other stuff to core. That web-server would also do a number of other things like the URL schema stuff that was discussed a while ago.

In order to do this the web-server would terminate the encryption (at least for the core stuff, for the authentication stuff we can do some magic to avoid this) and then simply send it as HTTP request to the ASP.NET stuff.

Therefor, every request would be a HTTP request which would be redirected to HTTPS creating an infinite loop. I actually have a Nginx configuration setup on my computer thus I already changed this.

I wanted to bring this to attention in a forum post but haven't gotten around doing that yet. If you want I can re-add it, it doesn't really matter for now.

misha130 replied:
Specifically in the case of the authentication project there will be no redirects for the URL schema (also the core project will not have it too).

The only thing that is needed is the HTTPS policy and since this donet gives support to all types of servers and not just nginx maybe it'll be best to put this back.

asynts replied:
https://discordapp.com/channels/634104110131445811/635510829499940912/682170974762041344

Maybe we should do all of this later. I mean use HTTP for development and worry about HTTPS later.

I have changed the authentication server to listen on HTTP only and changed the example client to listen to HTTP too. This would require a small change to core to work (replace `https://` in the authority with `http://`,) but it is so much simpler to run it.

cd70e34715ccb8eedd7e9a13bb12fe28141f2969

I would suggest that we create milestone called something like "Final Sprint" where we add all the issues that would only hinder development but have to be done for production. Then we create an issue "Use HTTPS for Production." and add it to this milestone. All the HTTPS policy stuff would go into that issue.

We would start working on that milestone when we get closer to release. I can think of a few more issues that belong into this category, like the URL schema stuff for example.

misha130 replied:
I have no problem with that but I think instead of the `NotImplemented` you can put `UseHsts `and `UseRedirectHttps` because no matter what the solution we will implement we will have to write this here.
Correct me if I am wrong?

asynts replied:
https://github.com/asynts/codidact-authentication/blob/bb77ebadd11626a40ee14357be1e0b6230a50ae1/src/WebApp/Startup.cs#L45-L52

You want me to replace that with the following?

~~~csharp
if (env.IsProduction())
{
    app.UseHsts();
    app.UseRedirectHttps();
}
else
{
    app.UseDeveloperExceptionPage();
}
~~~

Not sure if this is a good idea. There are a lot of things that have to be implemented before we can switch to production. Like `ISecretsService`, HTTPS, a web-server, *et cetera*. The exceptions sends the clear message, that you are not supposed to enable production yet.

asynts replied:
I have done it a bit different, now the 'Development' and 'Production' environments are allowed, choosing any other environment will throw an exception because that environment was not configured yet.

08252e49534558719edd51e3af56937ad2ad13cf

misha130 commented on src/WebApp/appsettings.json:
```diff
@@ -0,0 +1,25 @@
+                "ClientId": "codidact.com",
+                "ClientSecrets": [
+                    {
+                        "Value": "LCa0a2j/xo/5m0U8HTBBNBNCLXBkg7+g+YpeiGJm564="
```
Put a placeholder here

asynts replied:
That's the SHA256 for 'foo'. It looks like that because it's encoded in base64. (I think there is some metadata attached too.)

misha130 replied:
Well technically, security wise its not ok to keep this here. (I am not talking about foo, I am talking about the key/value)
I am not asking you to change this now but take a look at what I did on https://github.com/codidact/core/pull/60

https://github.com/codidact/core/blob/5b2b6c273e60e7e8b75b96d01aa56ade1c5642de/src/WebUI/Options/OpenIdConnectPostConfigureOptions.cs
https://github.com/codidact/core/blob/5b2b6c273e60e7e8b75b96d01aa56ade1c5642de/src/WebUI/Startup.cs

asynts replied:
86d35004bb522a17518d42566a3923884dd2c715

asynts replied:
@misha130 I got a different idea, we could simply create an `ISecretConfiguration` interface that is essentially an alias for `IConfiguration` and have it resolve a different configuration than the normal `IConfiguration`. In a development environment we could simply make it an alias for the normal `IConfiguration`:

f842cab2978f035f68fec0da0632b71be238b09b

The problem is that this removes the possibility for using asynchronous methods because the `IConfiguration` interface does not supply them. Do you expect that loading the secret configuration would be an expensive operation that, therefor, needs `async`/`await`?

misha130 replied:
ISecretService is not an alias for IConfiguration.
Only secrets should be taken from ISecretService.
This is why its async because it might be remote.

The whole issue came up when I tried storing the connection string in a dev secrets rep realizing I can't use it in production. So I created this to abstract what is used. In the future the secrets should be taken from the cloud or DB.

This is how I originally planned it at least.

asynts replied:
https://ptb.discordapp.com/channels/634104110131445811/635510829499940912/684078492773777514

misha130 commented on src/WebApp/appsettings.json:
```diff
@@ -0,0 +1,25 @@
+                    }
+                ],
+                "AllowedGrantTypes": [
+                    "client_credentials"
```
Why did you switch to client credentials over PKCE authorization code?
PKCE/Code is the recommended way.

I think you are exposing your client secret on every login request too?
Client Credentials is not supposed to be used between websites verification like this

asynts replied:
You are right. This should be `authorization_code`.

f930142c

misha130 commented on src/Infrastructure/Persistance/ApplicationDbContext.cs:
```diff
@@ -0,0 +1,17 @@
+    public class ApplicationDbContext : IdentityDbContext<ApplicationUser, IdentityRole<long>, long>
+    {
+        public ApplicationDbContext(DbContextOptions<ApplicationDbContext> options)
+            : base(options)
```
Where is the config files that renamed the table names from "AspNetUsers" to "users" etc?

asynts replied:
I changed the table names to match the pattern from the database schema discussions. I do not think they were formalized exactly but I tried to use the same naming convention.

These names are different from what [you originally suggested][2], is this okay?

d303a02870531a517b8313f372de004982a854ab

  [1]: https://github.com/asynts/codidact-authentication/tree/1/move-from-core.d/remove-table-prefixes

misha130 commented on tests/WebApp.Tests/OpenidConfigurationTests.cs:
```diff
@@ -0,0 +1,93 @@
+        [Fact]
+        public async Task RequestWithValidCredentials()
+        {
+            var client = new TokenClient(_http, new TokenClientOptions
```
These tests don't pass since you switched to auth_code

asynts replied:
The `appsettings.json` file is never read because I used `new HostBuilder()` instead of `WebHost.CreateDefaultBuilder()`:

https://github.com/asynts/codidact-authentication/blob/d303a02870531a517b8313f372de004982a854ab/tests/WebApp.Tests/OpenidConfigurationTests.cs#L24

The configuration comes from an in-memory collection where client credentials are configured:

https://github.com/asynts/codidact-authentication/blob/d303a02870531a517b8313f372de004982a854ab/tests/WebApp.Tests/OpenidConfigurationTests.cs#L31

asynts commented:
The original pull request used an in-memory database instead of Postgres like before. My reasoning was that during development, this would ensure a more predicable environment because there would be no remaining (possibly broken) database stuff from the last run.

I have now come to realize that this is very unpractical, everything must be manually be hard-coded somewhere to be preserved in the next run.

Therefor I switched to Sqlite instead. The advantage to Postgres is, that we can simply delete the database by deleting the file directly. Doing this with Postgres required running an SQL script which is very unpractical.

The idea is that we would switch to Postgres when we are closer to release or when we have to rely on database specific stuff. This would literally require changing three lines of code.

71859f22

asynts commented:
There is some configuration in core that has to be changed before it can be used with the authorization project. One problem was that you would get an SSL error if you use a self signed certificate for localhost. There might be a few options I missed.

I have created an example client that can be used as reference for core. We can remove it as soon as we have core configured correctly.

7ddd2182

asynts commented:
There is some configuration in core that has to be changed before it can be used with the authorization project. One problem was that you would get an SSL error if you use a self signed certificate for localhost. There might be a few options I missed.

I have created an example client that can be used as reference for core. We can remove it as soon as we have core configured correctly.

7ddd2182

asynts commented:
https://github.com/codidact/core/blob/c1e82e109b8ef0ea67cf94f3a04471660d186878/src/Infrastructure/Persistence/Configuration/IdentityConfiguration.cs

I actually searched everywhere for that code, could not find it. The reason is probably that I thought it belonged to the migration stuff and was automatically created.

In the [database schema forum post][1] the table names were singular, in that snippet they are plural.

Should I change the table names to be plural?

  > I am not sure about the Identity prefix because it implies [the] use of a specific library.

Is that a good thing?

  [1]: https://forum.codidact.org/t/db-schema-round-7/961?u=asynts

misha130 commented:
No, singular is correct. 
> Is that a good thing?
I suppose the benefit is you can clearly understand which tables are part of the library but a dba wouldn't likely understand the meaning here. In short, I can't tell if its good or bad

asynts commented:
I will leave it then. Changing table names is pretty much trivial, we could do this if someone has a good reason for it.

asynts commented:
I have had some issues with the return URLs. I do not know if they work correctly now, because a ton of different services interact with it and the documentation is really bad in this regard. I have created issue #4 for this, this pull request is already very large and should not be bloated with that too.

DoctaJonez commented:
> 1. In my mind, it's best to precisely specify what is included and what isn't, compared to a 300+ line `.gitignore` file where 299 lines are irrelevant. However, it's annoying to rewrite history because something was accidentally added. Should I keep the `.gitignore` file as is, or use a move greedy version instead?

What gitignore file are we using?  In my experience it has been best to use the standard gitignore file for the project type that you're using.  In this case, we should be using the [standard visual studio gitignore file](https://github.com/github/gitignore/blob/master/VisualStudio.gitignore).

asynts commented:
> What gitignore file are we using?

This pull request has one '.gitignore' file for each project. This is the one for the web application:

https://github.com/asynts/codidact-authentication/blob/c2366c0f5210888c9c99739ec41af79196d5eb7a/src/WebApp/.gitignore

I prefer this type of '.gitignore' file, because you can add a comment for every single thing explaining exactly why it is ignored. However, it is just a matter of time until someone accidentally commits a large binary file onto the master branch.

  > In this case, we should be using the standard visual studio gitignore file.

The core project currently uses the same '.gitignore' file you suggested, if we decide to use a more greedy version this would be the go to choice.

asynts commented:
I have moved a point from my to-do list into issue #5. I do not think that this has to be done now and would only add bloat to the pull request.

asynts commented:
What license should I put on the project? Core uses AGPL (which I totally disagree with) but @ArtOfCode-  said that we could do this differently for other projects.

https://forum.codidact.org/t/what-software-license-should-we-use/205/29?u=asynts

https://discordapp.com/channels/634104110131445811/634374616646811648/682215891269976086

misha130 commented on Codidact.Authentication.sln:
```diff
@@ -0,0 +1,116 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
+# Visual Studio 15
```
I see you are using Visual Studio 2017 (I think?) wouldn't this need to be ported to 2019?

asynts replied:
I created this file using `dotnet new sln` from the command line. From the documentation:

  >     # Visual Studio Version 16
  > The major version of Visual Studio that (most recently) saved this solution file. This information controls the version number in the solution icon. &mdash;&nbsp;[source][1]

I do not think anything needs to be changed here, the important part seems to be

    Microsoft Visual Studio Solution File, Format Version 12.00

Which seems to be the most recent version.

  [1]: https://docs.microsoft.com/en-us/visualstudio/extensibility/internals/solution-dot-sln-file?view=vs-2019&redirectedfrom=MSDN#definitions-1

misha130 commented on src/Client/Client.csproj:
```diff
@@ -0,0 +1,12 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
```
Why is there a client project even? Whats the point of this?
I looked at the pages/startup and couldn't figure it out.
Is this for testing?

Edit: If this is a further separation of responsibility I am not going to say anything but is it?
Edit2: if this is really just for testing to see if it works, remove it please. Make integration tests, not this. If anything this is an E2E test. Atleast move this to tests folder

asynts replied:
I added this because the project does not work with core without changing it's configuration.

The idea is that we have something that can be run without any complications. After codidact/core#61 is merged, I would add a pull request that changes the configuration to work with this project, and then remove this client entirely.

misha130 replied:
Well considet that the configuration changes that are needed to be done are client secret changes or authority changes which I would expect anyone who downloads this code and deploys to do on their own. These configuration changes should be noted in the readme like the connection string. 

Since we are removing this either way in the future or now decide on your own which you want to do

asynts replied:
46cd8cf632b649c460b5b4d25cec5a25f94cfd88
8c2b6cb65d6581a1bef4e9fff4d0f4ff89107373

misha130 commented on src/Infrastructure/DependencyInjection.cs:
```diff
@@ -0,0 +1,48 @@
+            services
+                .AddIdentityCore<ApplicationUser>(options =>
+                {
+                    options.Password.RequireDigit = false;
```
Note in the future we will need 1 unique, one lower one upper one digit.
For example: Aa123456!

asynts replied:
There was a [discussion about passwords][1]. My understanding was that we only check for the length of the password and accept the (possibly insecure) password they choose. I used eight as the minimum such that I could use 'password' for the test user...

  [1]: https://forum.codidact.org/t/mvp-account-database-scheme-suggestion/337/50?u=asynts

misha130 replied:
Alright if it was discussed then so shall it be

misha130 commented on src/Infrastructure/DependencyInjection.cs:
```diff
@@ -0,0 +1,48 @@
+                .AddInMemoryClients(configuration.GetSection("IdentityServer:Clients"))
+                .AddInMemoryIdentityResources(Config.GetIdentityResources())
+                .AddAspNetIdentity<ApplicationUser>()
+                .AddDeveloperSigningCredential();
```
If possible add `if(enviornment.isDevelopement)` here. Better yet open an issue for this if you want to.
The problem is passing the ENV variable to the this extension method because infrastructure refuses to recognize IHostEnv no matter the nugets I installed.

asynts replied:
bb77ebadd11626a40ee14357be1e0b6230a50ae1

I have created #6 to do this properly, but this will do for now.

misha130 commented on src/Infrastructure/Persistance/ApplicationDbContext.cs:
```diff
@@ -0,0 +1,30 @@
+            builder.Entity<IdentityUserLogin<long>>().ToTable("identity_user_login");
+            builder.Entity<IdentityUserClaim<long>>().ToTable("identity_user_claim");
+            builder.Entity<IdentityUserRole<long>>().ToTable("identity_user_role");
+            builder.Entity<IdentityUserToken<long>>().ToTable("identity_user_token");
```
For sake of organization, consider moving this code to a separate extension class.

asynts replied:
I have done it a bit different than in core and put the class into `Codidact.Authentication.Infrastructure.Persistence` instead of `Codidact.Authentication.Infrastructure.Persistence.Configuration`.

3b2cac174c2dbdcddce3e6798856b0062ae97122

misha130 commented on src/WebApp/Startup.cs:
```diff
@@ -0,0 +1,90 @@
+        }
+
+        private static void SeedDatabase(IApplicationBuilder app)
+        {
```
Sorry for asking about this 20 times but what is the reason for going back to this instead of opening a web page when there are no users?

asynts replied:
There has to be at least one user so that it is possible to test if the login works. That is why I added this code originally.

Then, some time ago, I setup a forum software for something else, and thought, wouldn't it be nice to have an interactive installer that sets the page title, creates an administrator, therefor, I created the `/account/firstuse` endpoint.

However, it really makes no sense to do this in the MVP. Setting up such a installer is not trivial and maintaining it would be a very annoying because you would really only use it once anyway.

Therefor, I created issue #3 and rolled back the changes.

misha130 commented on tests/WebApp.Tests/OpenidConfigurationTests.cs:
```diff
@@ -0,0 +1,85 @@
+        {
+            var client = new TokenClient(_http, new TokenClientOptions
+            {
+                Address = "http://localhost/connect/token",
```
Is the port correct here? Isn't it 5000 or 5001?

asynts replied:
This is a test server, as far as I know, there is no actual request made, therefor, no port. However, I could not find any documentation that actually confirms this.

https://github.com/asynts/codidact-authentication/blob/bb92f3574d8b1a26afbe405b3800a7b73a43ea58/tests/WebApp.Tests/OpenidConfigurationTests.cs#L44

asynts commented:
I think this pull request is ready to be merged, there are a few things that could get sorted out, but I would rather do them in separate pull requests, otherwise, this pull request would get to large.

asynts commented:
For the `.gitignore` question, I would like to point out, that the official Asp.NET Core project uses a very compact version too:

~~~none
# Folders
artifacts/
bin/
obj/
.dotnet/
.nuget/
.packages/
.tools/
.vs/
.vscode/
node_modules/
BenchmarkDotNet.Artifacts/
.gradle/
src/SignalR/clients/**/dist/
modules/

# File extensions
*.aps
*.binlog
*.dll
*.DS_Store
*.exe
*.idb
*.lib
*.log
*.pch
*.pdb
*.pidb
*.psess
*.res
*.snk
*.so
*.suo
*.tlog
*.user
*.userprefs
*.vspx

# Specific files, typically generated by tools
launchSettings.json
msbuild.ProjectImports.zip
StyleCop.Cache
UpgradeLog.htm
.idea
~~~

https://github.com/dotnet/aspnetcore/blob/a2568cbe1e8dd92d8a7976469100e564362f778e/.gitignore

So it might not be a bad idea after all...

asynts commented:
I moved all the files that are likely changed into `/docs/defaults/` and added instructions in `/docs/development.md`.

No more `git stash` when the `appsettings.json` have been changed.

b5626f78aad8b8ff6ae8633a4e800eba31652896

DoctaJonez commented:
> For the `.gitignore` question, I would like to point out, that the official Asp.NET Core project uses a very compact version too:
> ...
> So it might not be a bad idea after all...

Despite that, my preference would be to use a more comprehensive gitignore file to prevent mistakes.  A smaller file confers no advantages, it's not as if it will give us a performance benefit.  We're not in business of making gitignore files, and the list of metadata files that shouldn't be source controlled is only ever going to grow.  In my opinion, crafting our own is wasted effort when we have more pressing issues.

I guess we can revisit this issue when we have assets that are accidentally committed.

asynts commented:
> I guess we can revisit this issue when we have assets that are accidentally committed.

@DoctaJonez That does sound like a good compromise. We see if it works this way and change it otherwise.

asynts commented:
I have added a disclaimer to the `.gitignore` file.

be8ee1e68e781f525b5d604333b9445a2375bc5d

ArtOfCode- commented:
Is this ready to merge? If so, I'll merge both this and core#61.

asynts commented:
@ArtOfCode- Yes, this is ready to merge.
